File: built-in plugins dsp/mb-fm-lead.cpp
/**
 * MB FM Lead
 * Category : instrument
 * Type     : fm
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Cutting FM lead
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_FM_LEAD_H
#define MB_FM_LEAD_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbFmLead : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-fm-lead";
    static constexpr const char* PLUGIN_NAME    = "MB FM Lead";
    static constexpr const char* PLUGIN_TYPE    = "fm";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float modIndex = 7f;  // range [0, 20]
    float brightness = 0.8f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbFmLead() = default;
    ~MbFmLead() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.modIndex = std::clamp(params.modIndex, 0f, 20f);
        params.brightness = std::clamp(params.brightness, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB FM Lead
        return input;
    }
};

#endif // MB_FM_LEAD_H

File: built-in plugins dsp/mb-fm-lead.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbFmLeadPlugin: PluginDefinition = { id: 'mb-fm-lead', slug: 'mb-fm-lead', name: 'MB FM Lead', category: 'instrument', type: 'fm', version: '1.0.0', description: 'Cutting FM lead', author: 'Max Booster', oscillators: [{ type: 'sine', detune: 0, gain: 0.8 }], envelope: { attack: 0.01, decay: 0.3, sustain: 0.6, release: 0.3 }, parameters: [{ id: 'modIndex', name: 'Mod Index', type: 'float', defaultValue: 7, minValue: 0, maxValue: 20, automatable: true }, { id: 'brightness', name: 'Brightness', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true }, { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { modIndex: 7, brightness: 0.8, volume: 0.8 } };

export default MbFmLeadPlugin;
